clang/include/clang/Basic/DarwinSDKInfo.h

```diff
@@ -141,20 +143,55 @@ class DarwinSDKInfo {
     llvm::DenseMap<VersionTuple, VersionTuple> Mapping;
   };
 
+  // Special versions of hash and equal_to to map triples to system prefixes
+  // ignoring the Arch/SubArch.
+  struct TripleHash {
+    std::size_t operator()(const llvm::Triple &Val) const noexcept {
+      return hash_combine(Val.getVendor(), Val.getOS(), Val.getEnvironment(),
+                          Val.getObjectFormat());
+    }
+  };
+  struct TripleEqualTo {
+    constexpr bool operator()(const llvm::Triple &LHS,
+                              const llvm::Triple &RHS) const {
+      return (LHS.getVendor() == RHS.getVendor()) &&
+             (LHS.getOS() == RHS.getOS()) &&
+             (LHS.getEnvironment() == RHS.getEnvironment()) &&
+             (LHS.getObjectFormat() == RHS.getObjectFormat());
+    }
+  };
+
   DarwinSDKInfo(
       VersionTuple Version, VersionTuple MaximumDeploymentTarget,
-      llvm::Triple::OSType OS,
+      llvm::Triple Triple,
+      std::unordered_map<llvm::Triple, std::string, TripleHash, TripleEqualTo>
```

Comment:
`unordered_map` is generally discouraged in the programmers manual: https://llvm.org/docs/ProgrammersManual.html#other-map-like-container-options. Though I suspect the container size would always be generally small. Why did you pick this container? 

If you're ignoring arch, why not just hold on to the env & os type in a `std::pair` as the key? Wouldn't vendor always be `apple` when using `DarwinSDKInfo`?

Reply:
I did see that and initially I used `SmallDenseMap` but @felipepiovezan suggested that `unordered_map` was a better match since triple keys were a little big for what DenseMap is typically used for.

It's not just os + env but also object format. Which _should_ always be MachO I think, but if something ever changed in the future then we'd be stuck. Same for vendor. So I was thinking Triple was more future proof, though I guess `std::tuple` would work too.

Reply:
Yeah we'd need `operator <` for `map`, and there isn't really a logical implementation of that for triples so that's why I went with `unordered_map`. Is there any advantage of `map`? As far as I can tell it has the same downsides of `unordered_map` re allocations.

Maybe what would be best is a `SupportedTarget` struct that holds the triple pieces we want plus the system prefix, and then we could store those in a `SmallVector`. We could use that as the canonical platform representation, and still support looking them up by target triple.

Reply:
> Is there any advantage of map?

It's the same thing you said earlier, it's less code to achieve the same thing. 

> There isn't really a logical implementation of that for triples
I suspect that's because no one has ever needed to support ordering, but the usual way to do this is 
```
bool operator <(const auto &LHS, const auto &RHS) {
return std::tie(LHS.attr1, LHS.attr2) < 
       std::tie(RHS.attr1, RHS.attr2);
}
```

> Maybe what would be best is a SupportedTarget struct

SGTM.